### ISSUE TITLE
Allow usage of Set<T> inside a `for of` loop

### ISSUE THREAD

**Issue by shroudedcode** [Microsoft/TypeScript#21601]
**Search Terms:**

* typescript for of set

**Code:**

```ts
const strSet: Set<string> = new Set()
strSet.add('Hello')
strSet.add('World')
for (const str of strSet) {
  console.log(str)
}
```

[*Open in playground ...*](https://www.typescriptlang.org/play/#src=const%20strSet%3A%20Set%3Cstring%3E%20%3D%20new%20Set()%0D%0AstrSet.add('Hello')%0D%0AstrSet.add('World')%0D%0Afor%20(const%20str%20of%20strSet)%20%7B%0D%0A%20%20console.log(str)%0D%0A%7D)

**Expected behavior:**

* No error is shown.
* `str` is correctly inferred as `string`.

**Actual behavior:**

* A type error is shown.
![](https://user-images.githubusercontent.com/29176678/35769075-e6603978-0905-11e8-9c4e-ef57a6f9f5a6.png)
* `str` is incorrectly inferred as `any`.
![](https://user-images.githubusercontent.com/29176678/35769096-0d970e86-0906-11e8-925f-619155cfe7de.png)

> The `for...of` statement creates a loop iterating over iterable objects (**including** ⋯ **Set** ⋯).
> 
> *— [`for ... of` on MDN](https://developer.mozilla.org/en-US/docs/Web/JavaScript/Reference/Statements/for...of)*

**Comment by aluanhaddad** [comment]
This isn't a bug. For targets below es2015, a flag is required due to the complexity of the emit this necessitates for all iterable specific syntax regardless of the types involved.

Simply `--downlevelIteration` and go to town.

**Comment by aluanhaddad** [comment]
@shroudedcode it is true that there are a lot of issues and most were filed for `Map`. Also, there are some gotchas like needing to specify `--lib dom.iterable` to use iterables with types like `NodeList`.
